django/db/migrations/state.py

```diff
@@ -204,6 +239,50 @@ def alter_field(self, app_label, model_name, name, field, preserve_default):
             field.default = NOT_PROVIDED
         else:
             field = field
+        if self._relations is not None:
+            model_key = (app_label, model_name)
+            model_state = self.models[model_key]
+            old_remote_field = model_state.fields[name].remote_field
+            new_remote_field = field.remote_field
+            if old_remote_field and new_remote_field:
+                concretes, _ = self._get_concrete_models_mapping_and_proxy_models()
+                real_apps = set(self.real_apps)
+                if old_remote_field.model != new_remote_field.model:
+                    remote_model_key = resolve_relation(old_remote_field.model, *model_key)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self._relations[remote_model_key][model_key].remove((name, model_state.fields[name]))
+                    through = getattr(old_remote_field, 'through', None)
+                    if through:
+                        through_model_key = resolve_relation(through, *model_key)
+                        if through_model_key[0] not in real_apps and through_model_key in concretes:
+                            through_model_key = concretes[through_model_key]
+                        self._relations[through_model_key][model_key].remove((name, model_state.fields[name]))
+            else:
+                concretes, _ = self._get_concrete_models_mapping_and_proxy_models()
+                real_apps = set(self.real_apps)
+                if old_remote_field:
+                    remote_model_key = resolve_relation(old_remote_field.model, *model_key)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self._relations[remote_model_key][model_key].remove((name, model_state.fields[name]))
+                    through = getattr(old_remote_field, 'through', None)
+                    if through:
+                        through_model_key = resolve_relation(through, *model_key)
+                        if through_model_key[0] not in real_apps and through_model_key in concretes:
+                            through_model_key = concretes[through_model_key]
+                        self._relations[through_model_key][model_key].remove((name, model_state.fields[name]))
```

Comment:
It looks like the 10 lines above might be an exact copy of the same lines from the `if` clause above. Would it make sense to pull out chunks of common code into helper methods?

Reply:
You can reuse `resolve_model_field_relations()`.